### PR TITLE
doc: add vhost-user-blk documentation

### PR DESCRIPTION
**Pull request by kalyazin**
## Changes

Add documentation for the vhost-user-blk device.

## Reason

vhost-user-blk device is added in https://github.com/firecracker-microvm/firecracker/pull/4170 .

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- ~[ ] If a specific issue led to this PR, this PR closes the issue.~
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- ~[ ] API changes follow the [Runbook for Firecracker API changes][2].~
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- ~[ ] All added/changed functionality is tested.~
- ~[ ] New `TODO`s link to an issue.~
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: docs/api_requests/block-vhost-user.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+is a userspace protocol that allows to delegate Virtio queue processing
+to another userspace process on the host, as opposed to performing this task
+within Firecracker's VMM thread.
+
+In the vhost-user architecture, the VMM acts as a vhost-user frontend, and
+it is responsible for connecting to the backend via a Unix domain socket (UDS)
+and sharing sufficient information about the guest memory and Virtio queues.
+The vhost-user backend is supposed to receive all this information and perform
+handling of guests's requests. The UDS control socket is no longer used after
+the setup is done.
```
Comment (ShadowCurse) [inline docs/api_requests/block-vhost-user.md:19]: I think it is worth mentioning that it is not used in our implementation. Technically it can communicate back the config change (or maybe smth else as well).

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,10 @@
   `snapshot-editor` tool to print MicrovmState of vmstate snapshot file in
   a readable format. Also made the `vcpu-states` subcommand available on
   x86_64.
+- [#4170](https://github.com/firecracker-microvm/firecracker/pull/4170):
+  Added support for vhost-user block devices. Firecracker implements
```
Comment (ShadowCurse) [inline CHANGELOG.md:23]: It would be nice to include all related PRs here (at least known at this point).

File: docs/api_requests/block-vhost-user.md (hunk 1 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: fair, updated

File: CHANGELOG.md (hunk 2 above)
Comment (kalyazin) [reply CHANGELOG.md]: done

File: docs/api_requests/block-vhost-user.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
+# Vhost-user block device
+
+As an alternative to [block device](block-io-engine.md) `Sync` and `Async`
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:3]: nit:

```suggestion
As an alternative to file-backed [block device](block-io-engine.md) engine,
```

`Sync` and `Async` seem irrelevant here.

File: docs/api_requests/block-vhost-user.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+in general at [FOSDEM23](https://archive.fosdem.org/2023/schedule/event/sds_vhost_user_blk).
+
+[Vhost-user](https://qemu-project.gitlab.io/qemu/interop/vhost-user.html)
+is a userspace protocol that allows to delegate Virtio queue processing
+to another userspace process on the host, as opposed to performing this task
+within Firecracker's VMM thread.
+
+In the vhost-user architecture, the VMM acts as a vhost-user frontend, and
+it is responsible for connecting to the backend via a Unix domain socket (UDS)
+and sharing sufficient information about the guest memory and Virtio queues.
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:16]: It is also responsible for setting up the device, interacting with guest, feature negotiation and all, right?

File: docs/api_requests/block-vhost-user.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+The vhost-user backend is supposed to receive all this information and perform
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:17]:
```suggest
The vhost-user backend receives all this information and performs
```

File: docs/api_requests/block-vhost-user.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+handling of guests's requests. The UDS socket is only used for control purposes
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:18]: nit: `control purposes` -> `control plane purposes`.

File: docs/api_requests/block-vhost-user.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+Firecracker only implements a vhost-user frontend. Users are free
+to choose from [existing open source backends](#backends) or implement their
+own.
+
+## Topology
+
+Each vhost-user device connects to its own UDS socket. There is no way for
+multiple devices to share a single socket, as there is no way to differentiate
+messages related to devices.
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:29]: that is a design choice of the vhost-user protocol, right? If that is so, we could mention that explicitly.

File: docs/api_requests/block-vhost-user.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+There are two points when the vhost-user frontend communicates with
+the backend:
+
+1. Device initialisation. When a vhost-user device is created, Firecracker
+   connects to the corresponding UDS socket and negotiates Virtio and Vhost
+   features with backend and retrieves device configuration.
+1. Device activation. When the guest driver finishes setting up the device,
+   Firecracker shares memory tables and Virtio queue information with the backend.
+   As a part of this, Firecracker shares file descriptors for guest's memory
+   regions, as well as file descriptors for queue notifications.
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:45]: Just a thought (not necessarily actionable in this PR). A lot of this info is `vhost-user` generic, rather than `vhost-user-block` specific. Do you see value in having a separate doc for `vhost-user` generic info?

File: docs/api_requests/block-vhost-user.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+## Advantages
+
+While vhost-user block is considered an optimisation to Firecracker IO, a naive
+implementation of the backend is not going to improve performance.
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:50]: This reads like a disadvantage. I am not sure we need to mention that at all. Certainly, not here.

File: docs/api_requests/block-vhost-user.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+guest memory is created as a [memfd](https://man7.org/linux/man-pages/man2/memfd_create.2.html)
+and mapped as `MAP_SHARED`.
+
+#### File descriptor in procfs
+
+An open `memfd` is reflected in `procfs` as any other open file descriptor:
+
+```shell
+$ ls -l /proc/{pid}/fd | grep memfd
+lrwx------ 1 1234 1234 64 Nov  2 13:39 32 -> /memfd:guest_mem (deleted)
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:85]: Is the `(deleted)` part correct here?

File: docs/api_requests/block-vhost-user.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+At the moment, Firecracker does not close the `memfd`, because it must remain open
+until all the configured vhost-user devices have been activated and their info
+shared with the backends. This kind of tracking is not implemented in Firecracker,
+but may be implemented in the future. Meanwhile, users need to make sure that
+the access to the Firecracker's `procfs` tree is restricted to trusted processes
+on the host.
+
+On the backend side, it is advised that the backend closes
+the guest memory region file descriptors after mapping them into the address
+space.
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:100]: What effect does that have in the `memfd` file appearing in procfs? Also:

`mapping them into the address space.` -> `mapping them into its own address space`.

File: docs/api_requests/block-vhost-user.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+size. Since `memfd` that is used to back the guest memory is considered a file,
+the file size resource limit cannot be less than the biggest guest memory
+region. This does not require any special action from a user, but needs to be
+taken into consideration.
+
+### Remote code execution in the backend
+
+It is recommended to run Firecracker using the [jailer](../jailer.md). Since
+the vhost-user backend interacts with the guest via a Virtio queue, there is
+a potential for the guest to exercise issues in the backend code base
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:115]: nit: I _think_ `code base` -> `codebase`

File: docs/api_requests/block-vhost-user.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+to trigger undesired behaviours. Users should consider running their backend
+in a jailer or applying other adequate security measures to restrict it.
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:117]: maybe we should clarify that our jailer is not designed to jail arbitrary binaries, so users should use their own solutions (?)

File: docs/api_requests/block-vhost-user.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+
+### Rate limiting / cgroups
+
+Virtio block device in Firecracker has a [rate limiting capability](../design.md#io-storage-networking-and-rate-limiting).
+
+In the vhost-user case, Firecracker does not participate in handling requests
+from the guest, so rate limiting becomes backend's responsibility.
+
+As an additional indirect measure, users can make use of `cgroups` settings
+(either via Firecracker jailer or independently) in order to restrict host CPU
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:132]: I thought that Firecracker jailer cannot be used to run arbitrary binaries (see previous comment as well). Am I wrong?

File: docs/api_requests/block-vhost-user.md [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+consumption of the guest, which would transitively limit guest's IO activity.
+
+### Protection against defects in the backend code
+
+Due to potential defects in the backend (eg mislocating Virtio queues or writes
+to a wrong location in the guest memory), the guest execution may be affected.
+It is advised that customers monitor guest's health periodically.
+
+Additionally, in order to avoid orhpaned Firecracker processes if the backend
+crashes, the backend may need to send a `SIGBUS` signal to the Firecracker
```
Comment (bchalios) [inline docs/api_requests/block-vhost-user.md:142]: For my own information: Is `SIGBUS` a standard signal to send in these cases?

File: docs/api_requests/patch-block.md [hunk 16]
```diff
@@ -98,9 +100,41 @@ operation can be undone if the guest issues a write for the last sector of the
...
+
+Unlike with Virtio block device, with vhost-user block devices, Firecracker
+does not interact with the underlying block file directly (the vhost-user
+backend does). It means that changes to the file are not automatically seen
+by Firecracker. There is a mechanism in the
+[vhost-user protocol](https://qemu-project.gitlab.io/qemu/interop/vhost-user.html)
+for the backend to notify the frontend about changes in the device config
+via `VHOST_USER_BACKEND_CONFIG_CHANGE_MSG` message. This requires an extra
+UDS socket connection between the frontend and backend used for
+backend-originated messages. This mechanism _is not supported_ by Firecracker.
```
Comment (bchalios) [inline docs/api_requests/patch-block.md:120]: Did you want "is not supported" to be in italics or maybe bold?

File: docs/device-api.md [hunk 17]
```diff
@@ -14,118 +14,122 @@ a call to one of the [API Endpoints](#api-endpoints) will fail with a
...
-|                            | size                  |    O     |       O        |      O       |     **R**     |      O       |      O     |
-| `Vm`                       | state                 |    O     |       O        |      O       |       O       |      O       |      O     |
-| `Vsock`                    | guest_cid             |    O     |       O        |      O       |       O       |    **R**     |      O     |
-|                            | uds_path              |    O     |       O        |      O       |       O       |    **R**     |      O     |
-|                            | vsock_id              |    O     |       O        |      O       |       O       |    **R**     |      O     |
-| `EntropyDevice`            | rate_limiter          |    O     |       O        |      O       |       O       |      O       |    **R**   |
-
-<sup>\*</sup>: The `TokenBucket` can be configured with any combination of
+| Schema                     | Property              | keyboard | serial console | virtio-block | vhost-user-block |  virtio-net   | virtio-vsock | virtio-rng |
```
Comment (bchalios) [inline docs/device-api.md:39]: I honestly have no clue what value do we think this table has.

File: docs/api_requests/block-vhost-user.md (hunk 3 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: well, they are somewhat relevant because the page I'm referring to is named after engines. Updated.

File: docs/api_requests/block-vhost-user.md (hunk 4 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: updated

File: docs/api_requests/block-vhost-user.md (hunk 5 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: updated

File: docs/api_requests/block-vhost-user.md (hunk 6 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: done

File: docs/api_requests/block-vhost-user.md (hunk 7 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:37]: no, this is a property of the vhost-user protocol, as far as I know

File: docs/api_requests/block-vhost-user.md (hunk 13 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:125]: done

File: docs/api_requests/block-vhost-user.md (hunk 14 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:144]: this is talking about running the guest in the jailer to restrict its CPU time to indirectly influence the amount of IO

File: docs/api_requests/patch-block.md (hunk 16 above)
Comment (kalyazin) [reply docs/api_requests/patch-block.md]: updated

File: docs/device-api.md (hunk 17 above)
Comment (kalyazin) [reply docs/device-api.md:39]: agree, just updated them for consistency to illustrate that some of the properties are relevant for one of the device types

File: docs/api_requests/block-vhost-user.md (hunk 7 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:37]: that's what I meant. Maybe we should clarify that. But not extremely important.

File: docs/api_requests/block-vhost-user.md (hunk 14 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:144]: oh right, I see now! thanks for the clarification

File: docs/api_requests/block-vhost-user.md (hunk 8 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:53]: Thoughts about this? If you don't think we should do this here feel free to resolve.

File: docs/api_requests/block-vhost-user.md (hunk 9 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:58]: Any thoughts on this?

File: docs/api_requests/block-vhost-user.md (hunk 10 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:93]: If this is correct, feel free to resolve the comment.

File: docs/api_requests/block-vhost-user.md (hunk 11 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:108]: comments on this?

File: docs/api_requests/block-vhost-user.md (hunk 10 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:93]: yes, this is correct. I don't know for sure what it means, but I think it may be related to the fact the file is not visible at filesystem level.

File: docs/api_requests/block-vhost-user.md (hunk 11 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:108]: >What effect does that have in the memfd file appearing in procfs?

Does the following not answer your question?

>Any process on the host that has access to this file in `procfs` will be able
to map the file descriptor and observe runtime behaviour of the guest.

>mapping them into the address space. -> mapping them into its own address space.

I don't think it is essential, because it doesn't look possible to map something to another process's address space, does it?

File: docs/api_requests/block-vhost-user.md (hunk 7 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: done

File: docs/api_requests/block-vhost-user.md (hunk 11 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: >mapping them into the address space. -> mapping them into its own address space.

added, since I was at it anyway

File: docs/api_requests/block-vhost-user.md (hunk 15 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:154]: In that case, maybe we could rephrase this to something like:

`... if the backend crashes, it may need to send a signal, such as SIGBUS to Firecracker`

File: docs/api_requests/block-vhost-user.md (hunk 8 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md:53]: Works for me! I was mainly asking your view if that makes sense?

File: docs/api_requests/block-vhost-user.md (hunk 11 above)
Comment (bchalios) [reply docs/api_requests/block-vhost-user.md]: > Does the following not answer your question?

    Any process on the host that has access to this file in procfs will be able
    to map the file descriptor and observe runtime behaviour of the guest.

It does! Sorry, I missed that.

File: docs/api_requests/block-vhost-user.md (hunk 15 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: done, thanks

**Review by bchalios — DISMISSED** [review]
LGTM Thanks a lot for addressing the comments!

File: docs/api_requests/patch-block.md [hunk 18]
```diff
@@ -98,9 +100,41 @@ operation can be undone if the guest issues a write for the last sector of the
...
+by Firecracker. There is a mechanism in the
+[vhost-user protocol](https://qemu-project.gitlab.io/qemu/interop/vhost-user.html)
+for the backend to notify the frontend about changes in the device config
+via `VHOST_USER_BACKEND_CONFIG_CHANGE_MSG` message. This requires an extra
+UDS socket connection between the frontend and backend used for
+backend-originated messages. This mechanism **is not supported**
+by Firecracker. Instead, Firecracker makes use of the `PATCH /drives`
+API request to get notified about such changes. Such an API request
+only includes the required property (`drive_id`), because optional properties
+are not relevant to vhost-user.
```
Comment (ShadowCurse) [inline docs/api_requests/patch-block.md:124]: Should we mention PATCH here, considering it is not implemented yet? I think we can add this note in the PR that adds config notification. Plus it would be nice to add it to the vhost-user set of PRs in the CHANGELOG.

File: docs/api_requests/block-vhost-user.md [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+messages related to devices at the vhost-user protocol level.
+
+Each device can be served by a separate backend or a single backend can serve
+multiple devices.
+
+## Interactions with the backend
+
+There are two points when the vhost-user frontend communicates with
+the backend:
+
```
Comment (ShadowCurse) [inline docs/api_requests/block-vhost-user.md:46]: I saw you mention PATCH in the `patch-block.md`. In this case there are 3 places where FC communicates with backend.

File: docs/api_requests/patch-block.md (hunk 18 above)
Comment (kalyazin) [reply docs/api_requests/patch-block.md:124]: I have already mentioned the config change PR in the changelog. I don't think it's worth delaying this doc part, because we already know we must merge that other PR.

File: docs/api_requests/block-vhost-user.md (hunk 19 above)
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: That is a good point. Updated.

File: docs/api_requests/block-vhost-user.md [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+1. Device initialisation. When a vhost-user device is created, Firecracker
+   connects to the corresponding UDS socket and negotiates Virtio and Vhost
+   features with backend and retrieves device configuration.
+1. Device activation. When the guest driver finishes setting up the device,
+   Firecracker shares memory tables and Virtio queue information with the backend.
+   As a part of this, Firecracker shares file descriptors for guest's memory
+   regions, as well as file descriptors for queue notifications.
+1. Config update. When receving a [`PATCH` request](./patch-block.md#updating-vhost-user-block-devices-after-boot)
+   on a vhost-user backed drive, Firecracker rerequests the device config
+   from the backend in order to make the new config available to the guest.
```
Comment (ShadowCurse) [inline docs/api_requests/block-vhost-user.md:56]: I assume this should be 1. 2. 3. not 1. 1. 1.

Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:56]: no, markdown allows to do that and renders as 1, 2, 3. It's convenient in a way we don't need to update all numbers if we add something at the beginning.

Comment (ShadowCurse) [reply docs/api_requests/block-vhost-user.md:56]: oh ok, didn't know.

**Review by ShadowCurse — APPROVED** [review]

**Review by bchalios — APPROVED** [review]

File: CHANGELOG.md [hunk 21]
```diff
@@ -28,6 +28,13 @@
   E.g. the associated metrics for vhost-user block device with endpoint
   `"/drives/rootfs"` will be available under `"vhost_user_block_rootfs"`
   in the metrics json object.
+- [#4138](https://github.com/firecracker-microvm/firecracker/pull/4138),
+  [#4170](https://github.com/firecracker-microvm/firecracker/pull/4170),
+  [#4223](https://github.com/firecracker-microvm/firecracker/pull/4223),
+  [#4226](https://github.com/firecracker-microvm/firecracker/pull/4226):
```
Comment (sudanl0) [inline CHANGELOG.md:34]: we already have an entry for 4226, is this intentional to mention all changes together?

Comment (kalyazin) [reply CHANGELOG.md:34]: hm, indeed, looks like we mention it twice. Would it make sense to remove an entry about vhost-user-blk metrics and assume it a part of the implementation?

**Review by sudanl0 — APPROVED** [review]
CHANGELOG LGTM

**Review by xmarcalx — COMMENTED** [review]
Thanks for the documents. I have left some suggestions.

File: docs/api_requests/block-io-engine.md [hunk 22]
```diff
@@ -22,6 +22,9 @@ with the `io_engine` field taking two possible values:
 The `Sync` variant is the default, in order to provide backwards compatibility
 with older Firecracker versions.
 
+**Note** [vhost-user block device](./block-vhost-user.md) is another option
```
[the remark this answers is not in the document]
Comment (kalyazin) [reply docs/api_requests/block-io-engine.md:25]: I don't quite believe that it is necessary, because having another process is already a lot and clear that it requires a separate discussion in the doc. I just left it here to have a link to make users aware.

File: docs/api_requests/block-vhost-user.md [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+
+## Advantages
+
+While vhost-user block is considered an optimisation to Firecracker IO, a naive
+implementation of the backend is not going to improve performance.
+
+The major advantage of using a vhost-user device is the backend can implement
```
[the remark this answers is not in the document]
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: I believe both are dramatically correct, but added if that makes it more readable for someone.

File: CHANGELOG.md [hunk 24]
```diff
@@ -28,6 +28,13 @@
...
+  Added support for vhost-user block devices. Firecracker implements
+  a vhost-user frontend. Users are free to choose from existing open source
```
[the remark this answers is not in the document]
Comment (kalyazin) [reply CHANGELOG.md:36]: thanks, done

File: docs/api_requests/block-vhost-user.md [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+custom processing logic. It can use intelligent algorithms to serve block
+requests, eg by fetching the block device data over the network or using
```
[the remark this answers is not in the document]
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md]: done

File: docs/api_requests/block-vhost-user.md [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+## Backends
+
+There are a number of open source implementations of a vhost-user backend
+available for reference:
+
+1. [Qemu backend](https://github.com/qemu/qemu/tree/master/contrib/vhost-user-blk)
+1. [Cloud Hypervisor backend](https://github.com/cloud-hypervisor/cloud-hypervisor/tree/main/vhost_user_block)
+1. [crosvm backend](https://github.com/google/crosvm/blob/main/devices/src/virtio/vhost/user/device/block.rs)
+1. [SPDK backend](https://github.com/spdk/spdk/blob/master/lib/vhost/vhost_blk.c)
+
```
[the remark this answers is not in the document]
Comment (kalyazin) [reply docs/api_requests/block-vhost-user.md:80]: added

File: docs/api_requests/block-io-engine.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (xmarcalx) [reply docs/api_requests/block-io-engine.md:25]: ok thanks

File: CHANGELOG.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (xmarcalx) [reply CHANGELOG.md:36]: thanks

File: docs/api_requests/block-vhost-user.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (xmarcalx) [reply docs/api_requests/block-vhost-user.md]: thanks

File: docs/api_requests/block-vhost-user.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (xmarcalx) [reply docs/api_requests/block-vhost-user.md:80]: thanks

**Review by xmarcalx — APPROVED** [review]
LGTM thanks for the changes

**Review by ShadowCurse — APPROVED** [review]
